lib/sidekiq_publisher/client.rb

```diff
@@ -4,29 +4,77 @@
+    # those internal methods are now replicated here, in ezCater-maintained code, to help
+    # prevent future minor or patch updates to sidekiq from breaking this class.
+
     def bulk_push(items)
       payloads = items.map do |item|
         normed = normalize_item(item)
-        process_single(item["class"], normed) || nil
-      end.compact
+        result = middleware.invoke(item["class"], normed, normed["queue"], @redis_pool) { normed }
+        result || nil
+      end
+
+      raw_multi_push(payloads.compact)
+    end
+
+    private
 
+    # A copy of sidekiq's `raw_push` with the following changes:
+    #   * returns the number of payloads pushed
+    #   * uses `multi` instead of `pipelined` so that the entire batch is atomic together
+    #     (see: https://github.com/ezcater/sidekiq_publisher/pull/5#discussion_r188069548)
+    def raw_multi_push(payloads)
       pushed = 0
-      with_connection do |conn|
-        conn.multi do |transaction|
-          payloads.each do |payload|
-            atomic_push(transaction, [payload])
-            pushed += 1
+      @redis_pool.with do |conn|
+        # the retry/error-handling logic is copied from sidekiq 6.4:
+        retryable = true
+        begin
+          conn.multi do |transaction|
+            # TODO: right now, this implementation replicates the original (and current)
+            # behavior of sidekiq_publisher, which is to add each job to the transaction
+            # individually. However, this eliminates some of the advantage of sidekiq's
+            # native bulk API, which can add jobs as a single operation as long as they
+            # are all going to the same queue. A more-efficient implementation here might
+            # be to divide payloads into groups by their queue, and add each group to the
+            # transaction as a unit. That may be done in a future version of this gem.
+            payloads.each do |payload|
+              atomic_single_push(transaction, payload)
+              pushed += 1
+            end
           end
+        rescue Redis::BaseError => e
+          # 2550 Failover can cause the server to become a replica, need
+          # to disconnect and reopen the socket to get back to the primary.
+          # 4495 Use the same logic if we have a "Not enough replicas" error from the primary
+          # 4985 Use the same logic when a blocking command is force-unblocked
+          if retryable && e.message =~ /READONLY|NOREPLICAS|UNBLOCKED/
+            conn.disconnect!
+            retryable = false
+            retry
+          end
+          raise
         end
       end
-
       pushed
     end
 
-    private
-
-    def with_connection(&blk)
-      @redis_pool.with(&blk)
+    # a copy of sidekiq's `atomic_push` with the following changes:
+    #   * removed support for multiple payloads (accepts only one payload)
+    def atomic_single_push(conn, payload)
```

Comment:
The current implementation of Sidekiq's `atomic_push` may be found [here](https://github.com/mperham/sidekiq/blob/v6.4.2/lib/sidekiq/client.rb#L221) for reference.